[Filebeat] Improve ASA/FTD Ingest Pipeline

hitchatwork opened the pull request:
## What does this PR do?

<!-- Mandatory
-->
Fixes #21658



For messages 716002:
- Changed to GROK; allows for better parsing of event.reason
- Added field for cisco.webvpn.group_name
- Added field for event.reason per cisco docs for why session was terminated
- Added field for cisco.termination_user for the AAA username terminating the connection

For messages 722051:
- Add angle brackets to dissect to properly dissect the message, per cisco docs
- Added field for cisco.webvpn.group.name

For messages 305011:
- Change to GROK; allows for variance in message format with identity firewall

For messages 302020:
- Added GROK pattern to allows for variance in message format with identity firewall

For messages 302014/302016/302021:
- Added patterns and modified order of patterns of GROK to better match teardown messages
- Note that order of processing is important as the most specific messages are matched first, falling through to the appropriate match.
- Added temp fields for teardown initiator and user; defined in cisco docs but currently no real place to put them, but could be in future.
- Added icmp_type and icmp_code parsing for 302021 messages
- Changed duration matching from TIME to NOTSPACE, as long-lived connections (over 24 hours) don't match TIME.

And:

- Added descriptions to many fields to make them easier to find in 7.9+ Kibana Ingest Node Pipeline editor.
- Changed source.bytes field type from integer to long, since long-lived flows can surpass the capacity of an integer; ECS reference field is defined as long.
- Changed destination.bytes type from integer to long, since long-lived flows can surpass the capacity of an integer; ECS reference field is defined as long.

## Why is it important?

Improved parsing of ASA/FTD messages mentioned above.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
~- [ ] I have made corresponding changes to the documentation~
~- [ ] I have made corresponding change to the default configuration files~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

```
cd x-pack/filebeat
PYTEST_ADDOPTS="-k cisco" TESTING_FILEBEAT_FILESETS=asa,ftd mage -v pythonIntegTest
```

## Related issues

- Closes #21658

elasticmachine commented:
## :grey_exclamation: Build Aborted
  
  > The PR is not allowed to run in the CI yet
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-23766/detail/PR-23766/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-23766/detail/PR-23766/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-23766/detail/PR-23766/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-23766/detail/PR-23766/1//artifacts) 

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats


* Build Cause: `Pull request #23766 opened`


* Reason: `The PR is not allowed to run in the CI yet`


* * Start Time: 2021-01-29T19:30:46.749+0000


* Duration: 4 min 34 sec


* Commit: e1035606c770f57cde4c1875b87271164d3983c0




</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->







### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-23766/detail/PR-23766/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>

  
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-23766/runs/1/steps/83/log/?start=0">here</a></li>
<li>Description: <code>githubPrCheckApproved: The PR is not allowed to run in the CI yet. (Only users with write permissions can do so.)</code></l1>
</ul>
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2021-01-29T19:32:25.349Z]  > git remote # timeout=10
[2021-01-29T19:32:25.353Z]  > git config --get remote.origin.url # timeout=10
[2021-01-29T19:32:25.357Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2021-01-29T19:32:25.364Z]  > git merge 560dd4b2a70f655ab881901fa1dc30173087f090 # timeout=10
[2021-01-29T19:32:25.373Z]  > git rev-parse HEAD^{commit} # timeout=10
[2021-01-29T19:32:25.381Z]  > git config core.sparsecheckout # timeout=10
[2021-01-29T19:32:25.385Z]  > git checkout -f e1035606c770f57cde4c1875b87271164d3983c0 # timeout=15
[2021-01-29T19:32:29.475Z] Commit message: "Changed source.bytes and destination.bytes fields from integer to long, as they could exceed integer capacity."
[2021-01-29T19:32:29.475Z] First time build. Skipping changelog.
[2021-01-29T19:32:29.475Z] Cleaning workspace
[2021-01-29T19:32:29.634Z]  > git --version # timeout=10
[2021-01-29T19:32:29.637Z]  > git --version # 'git version 2.17.1'
[2021-01-29T19:32:29.641Z] fatal: bad object 48a15103e04816e78fb1bdcf6c8cb6ad37e6a8bd
[2021-01-29T19:32:29.476Z]  > git rev-parse --verify HEAD # timeout=10
[2021-01-29T19:32:29.480Z] Resetting working tree
[2021-01-29T19:32:29.480Z]  > git reset --hard # timeout=10
[2021-01-29T19:32:29.561Z]  > git clean -fdx # timeout=10
[2021-01-29T19:32:30.336Z] Timeout set to expire in 3 hr 0 min
[2021-01-29T19:32:30.360Z] The timestamps step is unnecessary when timestamps are enabled for all Pipeline builds.
[2021-01-29T19:32:30.584Z] [INFO] Number of builds to be searched 10
[2021-01-29T19:32:31.206Z] [INFO] 'shallow' is forced to be disabled when running on PullRequests
[2021-01-29T19:32:31.228Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-23766/src/github.com/elastic/beats
[2021-01-29T19:32:31.265Z] [INFO] gitCheckout: Checkout SCM PR-23766 with default customisation from the Item.
[2021-01-29T19:32:31.296Z] [INFO] Override default checkout
[2021-01-29T19:32:31.353Z] Sleeping for 10 sec
[2021-01-29T19:32:41.379Z] The recommended git tool is: git
[2021-01-29T19:32:41.423Z] using credential f6c7695a-671e-4f4f-a331-acdce44ff9ba
[2021-01-29T19:32:41.452Z] Wiping out workspace first.
[2021-01-29T19:32:41.460Z] Cloning the remote Git repository
[2021-01-29T19:32:41.460Z] Using shallow clone with depth 10
[2021-01-29T19:32:41.460Z] Avoid fetching tags
[2021-01-29T19:32:41.472Z] Cloning repository git@github.com:elastic/beats.git
[2021-01-29T19:32:41.495Z]  > git init /var/lib/jenkins/workspace/Beats_beats_PR-23766/src/github.com/elastic/beats # timeout=10
[2021-01-29T19:32:41.501Z] Fetching upstream changes from git@github.com:elastic/beats.git
[2021-01-29T19:32:41.502Z]  > git --version # timeout=10
[2021-01-29T19:32:41.505Z]  > git --version # 'git version 2.17.1'
[2021-01-29T19:32:41.506Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2021-01-29T19:32:41.510Z]  > git fetch --no-tags --progress -- git@github.com:elastic/beats.git +refs/heads/*:refs/remotes/origin/* # timeout=15
[2021-01-29T19:33:01.786Z] Cleaning workspace
[2021-01-29T19:33:01.800Z] Using shallow fetch with depth 10
[2021-01-29T19:33:01.800Z] Pruning obsolete local branches
[2021-01-29T19:33:01.768Z]  > git config remote.origin.url git@github.com:elastic/beats.git # timeout=10
[2021-01-29T19:33:01.772Z]  > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
[2021-01-29T19:33:01.781Z]  > git config remote.origin.url git@github.com:elastic/beats.git # timeout=10
[2021-01-29T19:33:01.789Z]  > git rev-parse --verify HEAD # timeout=10
[2021-01-29T19:33:01.792Z] No valid HEAD. Skipping the resetting
[2021-01-29T19:33:01.792Z]  > git clean -fdx # timeout=10
[2021-01-29T19:33:01.802Z] Fetching upstream changes from git@github.com:elastic/beats.git
[2021-01-29T19:33:01.803Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2021-01-29T19:33:01.807Z]  > git fetch --no-tags --progress --prune -- git@github.com:elastic/beats.git +refs/pull/23766/head:refs/remotes/origin/PR-23766 +refs/heads/master:refs/remotes/origin/master # timeout=15
[2021-01-29T19:33:03.610Z] Merging remotes/origin/master commit 560dd4b2a70f655ab881901fa1dc30173087f090 into PR head commit e1035606c770f57cde4c1875b87271164d3983c0
[2021-01-29T19:33:03.612Z]  > git config core.sparsecheckout # timeout=10
[2021-01-29T19:33:03.615Z]  > git checkout -f e1035606c770f57cde4c1875b87271164d3983c0 # timeout=15
[2021-01-29T19:33:05.139Z] Merge succeeded, producing e1035606c770f57cde4c1875b87271164d3983c0
[2021-01-29T19:33:05.139Z] Checking out Revision e1035606c770f57cde4c1875b87271164d3983c0 (PR-23766)
[2021-01-29T19:33:05.349Z] Commit message: "Changed source.bytes and destination.bytes fields from integer to long, as they could exceed integer capacity."
[2021-01-29T19:33:05.349Z] Cleaning workspace
[2021-01-29T19:33:05.619Z]  > git --version # timeout=10
[2021-01-29T19:33:05.623Z]  > git --version # 'git version 2.17.1'
[2021-01-29T19:33:05.635Z] fatal: bad object 48a15103e04816e78fb1bdcf6c8cb6ad37e6a8bd
[2021-01-29T19:33:06.363Z] Masking supported pattern matches of $GIT_USERNAME or $GIT_PASSWORD
[2021-01-29T19:33:06.988Z] + git fetch https://****:****@github.com/elastic/beats.git +refs/pull/*/head:refs/remotes/origin/pr/*
[2021-01-29T19:33:05.109Z]  > git remote # timeout=10
[2021-01-29T19:33:05.113Z]  > git config --get remote.origin.url # timeout=10
[2021-01-29T19:33:05.121Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2021-01-29T19:33:05.124Z]  > git merge 560dd4b2a70f655ab881901fa1dc30173087f090 # timeout=10
[2021-01-29T19:33:05.134Z]  > git rev-parse HEAD^{commit} # timeout=10
[2021-01-29T19:33:05.141Z]  > git config core.sparsecheckout # timeout=10
[2021-01-29T19:33:05.144Z]  > git checkout -f e1035606c770f57cde4c1875b87271164d3983c0 # timeout=15
[2021-01-29T19:33:05.350Z]  > git rev-parse --verify HEAD # timeout=10
[2021-01-29T19:33:05.356Z] Resetting working tree
[2021-01-29T19:33:05.357Z]  > git reset --hard # timeout=10
[2021-01-29T19:33:05.551Z]  > git clean -fdx # timeout=10
[2021-01-29T19:34:14.744Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-23766/src/github.com/elastic/beats/.git
[2021-01-29T19:34:14.777Z] Archiving artifacts
[2021-01-29T19:34:15.486Z] + git rev-parse HEAD
[2021-01-29T19:34:15.810Z] + git rev-parse HEAD
[2021-01-29T19:34:16.114Z] + git rev-parse origin/pr/23766
[2021-01-29T19:34:16.162Z] [INFO] githubEnv: Found Git Build Cause: pr
[2021-01-29T19:34:16.446Z] Masking supported pattern matches of $GITHUB_TOKEN
[2021-01-29T19:34:17.430Z] [WARN] githubApiCall: The REST API call https://api.github.com/repos/elastic/beats/pulls/23766/reviews return 0 elements
[2021-01-29T19:34:17.482Z] [INFO] githubPrCheckApproved: Title: [Filebeat] Improve ASA/FTD Ingest Pipeline - User: hitchatwork - Author Association: CONTRIBUTOR
[2021-01-29T19:34:17.783Z] ERROR: githubPrCheckApproved: The PR is not allowed to run in the CI yet
[2021-01-29T19:34:17.783Z] ERROR: githubPrCheckApproved: The PR is not allowed to run in the CI yet. (Only users with write permissions can do so.)
[2021-01-29T19:34:17.855Z] [INFO] Let's stop build #1. The PR is not allowed to run in the CI yet
[2021-01-29T19:34:17.877Z] Sleeping for 5 sec
[2021-01-29T19:34:18.885Z] Stage "Lint" skipped due to earlier failure(s)
[2021-01-29T19:34:18.951Z] Stage "Build&Test" skipped due to earlier failure(s)
[2021-01-29T19:34:19.019Z] Stage "Packaging" skipped due to earlier failure(s)
[2021-01-29T19:34:19.105Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-23766/src/github.com/elastic/beats
[2021-01-29T19:34:19.597Z] Running on Jenkins in /var/lib/jenkins/workspace/Beats_beats_PR-23766
[2021-01-29T19:34:19.711Z] [INFO] getVaultSecret: Getting secrets
[2021-01-29T19:34:19.788Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-01-29T19:34:20.598Z] + chmod 755 generate-build-data.sh
[2021-01-29T19:34:20.598Z] + ./generate-build-data.sh https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-23766/ https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-23766/runs/1 ABORTED 213582
[2021-01-29T19:34:21.148Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-23766/runs/1/steps/?limit=10000 -o steps-info.json
[2021-01-29T19:34:21.399Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-23766/runs/1/tests/?status=FAILED -o tests-errors.json
[2021-01-29T19:34:21.399Z] Retry 1/3 exited 22, retrying in 1 seconds...
[2021-01-29T19:34:22.742Z] Retry 2/3 exited 22, retrying in 2 seconds...
[2021-01-29T19:34:24.603Z] Retry 3/3 exited 22, no more retries left.

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
Pinging @elastic/siem (Team:SIEM)

elasticmachine commented:
Pinging @elastic/security-external-integrations (Team:Security-External Integrations)

P1llus commented on x-pack/filebeat/module/cisco/asa/test/additional_messages.log-expected.json:
```diff
@@ -838,7 +850,7 @@
         "log.level": "informational",
         "log.offset": 2298,
-        "network.bytes": 0,
+        "network.bytes": "0",
```
It seems that networ.bytes was moved from long to string, might need to convert these

hitchatwork replied:
Just committed a fix for this (added conversion for network.bytes in the conversion section near end of pipeline).

P1llus commented on x-pack/filebeat/module/cisco/asa/test/asa-fix.log-expected.json:
```diff
@@ -19,6 +19,7 @@
         "event.kind": "event",
         "event.module": "cisco",
         "event.original": "%ASA-6-302016: Teardown UDP connection 110577675 for Outside:10.123.123.123/53723(LOCAL\\Elastic) to Inside:10.233.123.123/53 duration 0:00:00 bytes 148 (zzzzzz)",
+        "event.reason": "(zzzzzz)",
```
Is this just because of testdata or expected to be a reason?

hitchatwork replied:
Just commited a fix for this - one of the patterns in the list was out of order.

P1llus commented on x-pack/filebeat/module/cisco/meraki/test/generated.log-expected.json:
```diff
@@ -60,8 +60,8 @@
         "related.ip": [
-            "10.15.16.212",
-            "10.102.218.31"
+            "10.102.218.31",
```
I would recommend when generating test data that you specify the module and fileset, I get this issue from time to time as well.
Depending on how you run your expected commands it should be something like:
`TESTING_FILEBEAT_MODULES=cisco TESTING_FILEBEAT_FILESETS=asa`

hitchatwork replied:
Good to know...I had been using:
GENERATE=1 TESTING_FILEBEAT_MODULES=cisco mage pythonIntegTest

Which helps explain why it was updating the meraki stuff even though I hadn't changed anything having to do with Meraki.

P1llus reviewed (commented):
Adding some small comments

jsoriano commented on x-pack/filebeat/module/cisco/asa/test/additional_messages.log-expected.json:
```diff
@@ -141,7 +143,8 @@
         "log.level": "informational",
         "log.offset": 324,
         "network.direction": "inbound",
-        "network.protocol": "icmp",
+        "network.iana_number": 1,
+        "network.transport": "icmp",
```
Moving this field from `network.protocol` to `network.transport` can be a breaking change. Is this expected?

@P1llus wdyt?

hitchatwork replied:
Not sure if it is a breaking change. I was parsing according to ECS field reference, ICMP is better mapped to network.transport since it has it's own IANA number.  Also, it's not technically an L7 protocol like HTTP or FTP, which is how network.protocol is defined in ECS.

P1llus replied:
Yeah you are correct on that @hitchatwork , it's not that your mapping was wrong, it's just that if the current mapping was indeed incorrect, fixing it would mean that other users that might already be using the field, it would break for them.
Might need @andrewkroh or @adriansr to pitch in, in terms of how we want to apply this.

hitchatwork replied:
@P1llus - I could change it so that it matches what is currently set, so this doesn't sit in review purgatory.  After this is reviewed and if commited, I can open a separate PR to "fix" the classification of ICMP from protocol to transport.  LMK.  Thanks!

P1llus replied:
I think taking it in two separate PRs would make it easier to push the rest of the changes, though just wanted to see if anyone else would pitch in as well

hitchatwork replied:
@P1llus - I just created a 'set' clause if the network.tranport was set to 'icmp', then set the network.protocol to be 'icmp'.  That should prevent any breaking changes and help improve the consistency of nomenclature with ECS.

hitchatwork replied:
I reset the ICMP fields to what they were originally.  Will create a separate PR for fixing ICMP classification as transport instead of protocol.

jsoriano commented:
jenkins run the tests please

elasticmachine commented:
Pinging @elastic/ingest-management (Team:Ingest Management)

hitchatwork commented:
@jsoriano - could you kick jenkins off again?  I made adjustments from @P1llus comments and also updated from master. Thanks!

andrewkroh commented:
jenkins, run tests

hitchatwork commented:
Thanks for kicking that off @andrewkroh - 4 of the tests failed (dictionary item adds due to ICMP processing).  I decided to simplify the PR by removing any changes to ICMP at all (will create a seperate PR for that).  Could you kick Jenkins off again?

andrewkroh commented:
jenkins, run tests

hitchatwork commented:
I'm not sure why this build failed.  Looks like something maybe in the backend with the Windows 2019 environment setup for the build?

andrewkroh commented:
jenkins, run tests

hitchatwork commented:
Thanks a lot @andrewkroh - looks like it passed and can be reviewed by the relevant teams.  Appreciate the assistance.

andrewkroh reviewed (changes requested):
Thanks for your work on this. It LGTM, but it needs an entry in the changelog to describe the change. Can you please put a note into CHANGELOG.next.asciidoc at the root of the repo.

hitchatwork commented:
> Thanks for your work on this. It LGTM, but it needs an entry in the changelog to describe the change. Can you please put a note into CHANGELOG.next.asciidoc at the root of the repo.

Added changelog entry.

andrewkroh commented:
Looks like we have a conflict in fields.go due to the fields changes. This will need a rebase.

hitchatwork commented:
> Looks like we have a conflict in fields.go due to the fields changes. This will need a rebase.

@andrewkroh - I have rebased and cleaned up some of the commits.  LMK if something else needs work.  Thanks!

andrewkroh commented:
jenkins, run tests

hitchatwork commented:
@andrewkroh if you could take a look, seems it's stuck in build or something odd.  Thanks!

andrewkroh commented:
jenkins, run tests

hitchatwork commented:
@andrewkroh looks like the build is failing from when you pushed in the golden files commit.  Looks like expected output for cisco AMP module changed, and that's failing the build.

andrewkroh commented:
jenkins, run tests
